Panos network operations base

adambaumeister opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
@GuyAfik 

This is a PR that includes only the **pan-os-platform** series of commands (17 commands). It also removes the decorators, but maintains the dataclasses. Otherwise, the code is unchanged from the previous PR; https://github.com/demisto/content/pull/17430

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [X] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+        return all_devices
+
+    @classmethod
+    def build_from_string(cls, hostnames: str, username: str, password: str, api_key: str = None):
```
why don't you just build it from the constructor of ```Topology``` class instead of using a ```classmethod``` decorator?

adambaumeister replied:
when I wrote it I was thinking I might have other ways of building the topology, so it makes sense to split them into class methods for example `build_from_string`, `build_from_device`, `build_from_integration_context`.

GuyAfik replied:
Very nice thinking, I like your idea, keep those class methods. 
In addition for every method of that kind, it would be nice if you can add a doc string with an example explaining that.
For example what does build from string means? and what does build from device mean? I would be happy to understand the context better.

Overall incredible work, well done!

adambaumeister replied:
Updated this docstring slightly to provide this context

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+                    serial: device
+                }
+
+    def firewalls(self, filter_string: str = None) -> Generator[Firewall, None, None]:
```
why returning here a tuple where its last two values are ```None```?

adambaumeister replied:
Fixed by using Iterator instead

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+        for firewall in firewall_objects.values():
+            yield firewall
+
+    def all(self, filter_string: str = None) -> Generator[Union[Firewall, Panorama], None, None]:
```
why returning here a tuple where its last two values are ```None```?

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+            for value in self.firewall_objects.values():
+                yield value
+
+    def active_devices(self, filter_str: str = None) -> Generator[Union[Firewall, Panorama], None, None]:
```
why returning here a tuple where its last two values are ```None```?

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+
+        raise TypeError(f"{type(device)} is not valid as a topology object.")
+
+    def top_level_devices(self) -> Generator[Union[Firewall, Panorama], None, None]:
```
why returning here a tuple where its last two values are ```None```?

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+        self._params = tuple(params)
+
+
+class URLFilteringProfile(VersionedPanObject):
```
please elaborate what is the purpose of these classes inheriting from VersionedPanObject

adambaumeister replied:
These classes are used directly by the pan-os-python package. In some cases, pan-os-python does not implement an object model for the type of configuration we need to look for. Instead of falling back to the raw XML api,  we can simply write the models ourselves, and as long as we inherit from VersionedPanObject, these can be used in the exact same way any other object can.

In the long term we would contribute these models back to the pan-os-python project so we could pull them out of this code.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+
+        # params
+        params = []
+        params.append(
```
use here a for loop

GuyAfik commented:
Hi @adambaumeister, Thank you for the amazing work you have done. a few points:

1) Make sure that you add the commands to the README file of the integration. You can use the ```desmito-sdk generate-docs``` command to help you with that. if you need my help i'd be glad to help.
2) Please bump up and update the release notes according to standards, if you need my help I would be glad to help.
3) My suggestion is to try to avoid inferring the type of the variable when you create it, makes the code more readable in my opinion.
4) add doc-strings for all the functions that you created so for other developers it would be more easy to understand what it does.  
5) I liked a lot that you defined a class that explains if the commands there refer to panorama/firewall, well done!
6) if you can explain to me what is the benefit of you using ```dataclass``` it would be great.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -4564,58 +4580,322 @@ script:
+    outputs:
+    - contextPath: PANOS.ShowArp.Summary.hostid
+      type: Unknown
+    - contextPath: PANOS.ShowArp.Summary.max
```
I don't think this is a good idea - we don't need to change the existing context outputs but the new commands shouldn't be named "Panorama" as it's a misleading name. It's somewhat nice to seperate them anyway I think?

GuyAfik replied:
From my point of you I don't see a problem, @bakatzir what's your input on that matter?

bakatzir replied:
We are not changing the Output paths of already existing commands.
For the new commands that Adam is adding, as this is new content, and the proper product name is `PAN-OS`, and not `Paranorama`, I concur with Adam on this matter.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -4564,58 +4580,322 @@ script:
+    - contextPath: PANOS.TemplateStackOp.name
+      description: Template Stack Name
+      type: Unknown
+  dockerimage: demisto/pan-os-python
```
The docker image PR is still pending; https://github.com/demisto/dockerfiles/pull/6906

GuyAfik replied:
Good work, you have some conflicts [here](https://github.com/demisto/dockerfiles/pull/6906), I believe you should fix it

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -1,16 +1,43 @@
+import panos.errors  # type: ignore
+
+# -- This is a way to get around trimming commonserverpython on import
+try:
```
This makes it possible to import the code with external tools like demisto-sdk, for the purpose of code generation. It's not explicitly nessecary.

GuyAfik replied:
if it's not explicitly necessary, lets avoid that.

adambaumeister replied:
Done

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -1,16 +1,43 @@
+    pass
+
+from xml.etree.ElementTree import Element
+from panos.base import PanDevice, VersionedPanObject, Root, ENTRY, VersionedParamPath  # type: ignore
```
I can't remember the specific purpose on this one, I think something about unused imports when the imported object is referenced but not instantiated (i.e this is incorrect MyPy handling).

adambaumeister replied:
Done

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -1,16 +1,43 @@
 
 # disable insecure warnings
-requests.packages.urllib3.disable_warnings()
+requests.packages.urllib3.disable_warnings()  # type: ignore
```
The original line in the code was generating mypy warnings, so I silenced it as it's been there for a very long time without being fixed.

adambaumeister replied:
Done - will see if build fails..

adambaumeister replied:
Fixed this, for real this time!

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -6996,7 +7138,8 @@ def panorama_install_file_content_update_command(args: dict):
             'Status': 'Pending'
         }
         entry_context = {"Panorama.Content.Install(val.JobID == obj.JobID)": content_install_info}
-        human_readable = tableToMarkdown('Result:', content_install_info, ['JobID', 'Status'], removeNull=True)
```
This is my fault - my IDE was using a different max line length and I was using it to auto-format the code which cut these commands up. Fixed.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -4627,4 +4907,3 @@ script:
 tests:
 - palo_alto_firewall_test_pb
 - palo_alto_panorama_test_pb
-fromversion: 5.0.0
```
I didn't - demisto-sdk format might have?

Fixed by re-adding.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+
+        # params
+        params = []
+        params.append(
```
See above comment

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -1758,8 +1758,8 @@ script:
       default: false
-      description: The element to change (“url”, “recurring”, “certificate_profile”,
-        “description”).
+      description: "The element to change (\u201Curl\u201D, \u201Crecurring\u201D\
```
Fixed - I have no idea what happened here.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -4564,58 +4580,322 @@ script:
+      required: true
+    description: Installs specific content update file.
+    name: panorama-install-file-content-update
     outputs:
```
All of these have been reverted - I think what happened here is that the YAML got reformatted at some point by PyYAML.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -4614,8 +4614,286 @@ script:
+    outputs:
+    - contextPath: PANOS.ShowRouteSummary.Summary.hostid
+      description: ID of PAN-OS host
+      type: number
```
I am sorry maybe I misled you, but those need to be case sensitive. These are the options: 
![image](https://user-images.githubusercontent.com/53861351/154674920-9b0ae279-9c71-4cd6-b68e-ed6300722f1e.png)

Please change that for all the context outputs types

adambaumeister replied:
Done

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/README.md:
```diff
@@ -4687,3 +4687,558 @@ Installs specific content update file.
+
+| **Path** | **Type** | **Description** |
+| --- | --- | --- |
+| PANOS.ShowArp.Summary.hostid | string | ID of PAN-OS host | 
```
Very nice work on the README! Make sure that when you update ```string``` into ```String``` that you update the README as well

adambaumeister replied:
Done

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/README.md:
```diff
@@ -4687,3 +4687,558 @@ Installs specific content update file.
+                {
+                    "hostid": "0111112222333444",
+                    "interface": "ethernet1/1",
+                    "ip": "192.168.1.1",
```
make sure there isn't any sensitive data that will be part of the README such as IPs/fqdn and so on.

adambaumeister replied:
I slightly munged some of these IPs - they are all internal to my lab so no risk of secrets in there, but its a bit easier to add them to .secrets-ignore this way.

GuyAfik commented on Packs/PAN-OS/pack_metadata.json:
```diff
@@ -2,7 +2,7 @@
     "description": "Manage Palo Alto Networks Firewall and Panorama. For more information see Panorama documentation.",
     "support": "xsoar",
-    "currentVersion": "1.6.19",
+    "currentVersion": "1.6.20",
```
lets bump it in a ```minor``` bump instead of ```revision``` as this is quite a big change.

adambaumeister replied:
Would that be 1.7.1 or 1.7.0?

GuyAfik replied:
it should be 1.7.0 after solving the conflicts

GuyAfik replied:
Well, sorry I misled you, apparently this PR wasn't updated against the master for a long time.
We already have 1.8.8 at our master version, so 1.9.0 is good.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7043,2029 @@ def panorama_install_file_content_update_command(args: dict):
+        self._params = tuple(params)
+
+
+def run_op_command(device: Union[Panorama, Firewall], cmd: str, **kwargs) -> Element:
```
what is an ```op``` command?

adambaumeister replied:
Op = operational, the PANOS API refers to it as an "OP" command

GuyAfik commented:
Hi @adambaumeister, incredible work so far! Added a few more comments regarding README and Release-notes. 

**make sure to fix the conflicts issues as it prevents from some of the tasks of the build to run**.

If you need my help at any point with anything feel free to reach out to me.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+    result: Element = device.op(cmd, **kwargs)
+    if "status" in result:
+        if result.attrib.get("status") != "success":
+            raise OpCommandError
```
Done

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+
+        raise TypeError(f"{type(device)} is not valid as a topology object.")
+
+    def top_level_devices(self) -> Generator[Union[Firewall, Panorama], None, None]:
```
this is a Generator, so the types correspond to YieldType, SendType and ReturnType. In this case, the function only *yields* objects of type Firewall/Panorama, so the other values are None.

GuyAfik replied:
I suggest to remove that type annotation as it can be confusing

adambaumeister replied:
I replaced all instances of `Generator` with `Iterator` instead. This should make it much clearer.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+
+        raise TypeError(f"{type(device)} is not valid as a topology object.")
+
+    def top_level_devices(self) -> Generator[Union[Firewall, Panorama], None, None]:
```
This function is kind of an interface - Firewall And Panorama objects both act very similar so it's purpose is to yield any top level device regadless of type.

GuyAfik replied:
so maybe you can keep that function and split the code inside into two functions as mentioned above. WDYT?

adambaumeister replied:
I split this up slightly - there's no benefit i can see for having two top_level_devices functions as the logic is very simple (it's literally "is there panorama devices? ok, then return those. Otherwise, return the firewall devices". Having top_level_firewall_devices would never be used.

In the future we may expand this logic to include device-groups and templates, where we would have different functions for things like `get_devices_by_top_level_device_group(device_group)` but that's not required now.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+
+    def active_top_level_devices(self, device_filter_string: str = None):
+        """
+        Yields active top level devices
```
Fixed

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+                filter_str: devices.get(filter_str)
+            }
+
+        device: PanDevice
```
When we want a type hint for a variable in a for loop, we must define it before the foor loop. See here; https://www.python.org/dev/peps/pep-0526/#id11

GuyAfik replied:
as I said before, I suggest not defining the variables with type hints. 
The type hinting should only be in the function arguments / what the function returns.

adambaumeister replied:
ahh yes, good point, we can move it to the argument instead by using the type i.e `Dict[str, PanDevice]`. Fixed

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+                    serial: device
+                }
+
+    def firewalls(self, filter_string: str = None) -> Generator[Firewall, None, None]:
```
See above, comment - this is a generator type

GuyAfik replied:
I suggest to remove that type annotation as it can be confusing and maybe add a doc string to the function that explains that

adambaumeister replied:
Fixed by changing to iterator

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+        for firewall in firewall_objects.values():
+            yield firewall
+
+    def all(self, filter_string: str = None) -> Generator[Union[Firewall, Panorama], None, None]:
```
See above, comment - this is a generator type

GuyAfik replied:
I suggest to remove that type annotation as it can be confusing and maybe add a doc string to the function that explains that

adambaumeister replied:
Fixed by changing to iterator

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7025,6 +9161,8 @@ def main():
         if demisto.command() == 'test-module':
             panorama_test()
+            # Test the topology is working as well
+            run_test_command(test_topology)
```
done!

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7377,6 +9515,20 @@ def main():
+        # The following code exists PURELY to supress XSOAR linter errors. This code is
+        # unreachable as these commands are routed via the COMMAND object.
+        elif demisto.command() == 'pan-os-platform-get-arp-tables':
+            run_normal_command(get_arp_tables, demisto.args())
```
I've changed this to make it more simplistic. We still have a function for converting the dataclasses into CommandResults, but otherwise it's a more procedural flow of code.

GuyAfik replied:
awesome work Adam!!!

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7387,5 +9539,5 @@ def main():
 
 
-if __name__ in ["__builtin__", "builtins", '__main__']:
+if __name__ in ('__main__', '__builtin__', 'builtins'):
```
I guess demisto-sdk format? very strange. Updated.

GuyAfik replied:
might very well be

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+    )
+
+
+def run_test_command(func: Callable):
```
Done

GuyAfik replied:
I still don't see the change

adambaumeister replied:
I changed the test function but I still seperated it next to the existing panorama_test command. I think it makes sense to test both aspects of this integration seperately, as puttinig the topology test under panorama_test function is misleading.

GuyAfik replied:
this could be problematic, basically we should have a function that checks both of them and only if both of these functions return 'ok', then the test module should succeed in my opinion? could there be a possibility where we won't need to test_topology function while testing? if yes lets discuss that one

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+                        peer=peer_serial
+                    ))
+
+        if len(result) == 1:
```
So this is a good question and it's due to the way XSOAR handles list results. It seems like if you run a command twice, and each time it returns a 1-item list, this will turn into two output values in the context, even if the output_key_field is the same.

So the check here is, if its' a 1 item list, simply return that item, so when it goes into the context we don't create duplicate values.

GuyAfik replied:
that's interesting, would you mind to take a look at this together with me? lets discuss this further

adambaumeister replied:
Let me know if you want to jump on a quick call for this one

GuyAfik replied:
yeah lets jump into a call this week.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+
+        # The below is very important for XSOAR to de-duplicate the returned key. If there is only one obj
+        # being returned, return it as a dict instead of a list.
+        if len(result_data) == 1:
```
See above

GuyAfik replied:
that's interesting, would you mind to take a look at this together with me? lets discuss this further

adambaumeister replied:
sure, let me know when you want to jump on a call

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+            )
+
+        show_system_info = UniversalCommand.get_system_info(topology, hostid)
+        show_system_info_result = show_system_info.result_data[0]
```
No - system info will always have at least one item in it unless it fails, in which case it will raise.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+        Serializes the topology information for storage (for example, in the incident context)
+        :return: dict of topology information
+        """
+        frozen_topology = FrozenTopology
```
as I said before, I suggest not defining the variables with type hints.
The type hinting should only be in the function arguments / what the function returns.

adambaumeister replied:
I simply removed this function. It's not needed.

adambaumeister commented:
I updated to fix all these other issues, and I resolved the conflicts via a merge. Please double check the merge worked correctly.

I can't get the build to succeed though - there are several MyPy errors I'm not sure how to resolve, if you can help?

I also added new release notes, 1.9.0, as that's the current version? update-release-notes did not work for me here, I had to add them manually for some reason.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+
+    def get_by_filter_str(self, filter_string: str = None) -> dict:
+        """Filters all devices and returns a dictionary of matching."""
+        all_devices = {**self.firewall_objects, **self.panorama_objects}
+        all_devices = Topology.filter_devices(all_devices, filter_string)
+        return all_devices
```
```suggestion
        return Topology.filter_devices({**self.firewall_objects, **self.panorama_objects}, filter_string)
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+                        api_password=password,
+                    )
+                # Set the timeout
+                device.timeout = 120
```
maybe define here a class member for the timeout variable

adambaumeister replied:
I added this as a globalvar - in a subsequent PR, as we introduce more commands, we should make this a parameter. Not now though, because it would incorrectly cause people to assume that timeout value is honored by the entire integration.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+            # If it's already a direct connection
+            return firewall
+
+        ip_address = firewall.show_system_info().get("system").get("ip-address")
```
```suggestion
        ip_address = (firewall.show_system_info().get("system") or {}).get("ip-address")
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+        Given a device, returns all the possible configuration containers that can contain objects -
+        vsys, device-groups, templates and template-stacks.
+        """
+        containers: List[Union[tuple[PanDevice, Callable], tuple[PanDevice, PanDevice]]] = []
```
```suggestion
        containers = []
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+        return_containers = []
+
+        if container_name:
+            for container in containers:
```
the ```containers``` is a tuple of two values?

adambaumeister replied:
containers is a list of tuple. The list itself can be variable length

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    outgoing_advertised: int = 0
+
+    def __post_init__(self):
+        self.incoming_total = int(self.incoming_total)
+        self.incoming_accepted = int(self.incoming_accepted)
+        self.incoming_rejected = int(self.incoming_rejected)
+        self.policy_rejected = int(self.policy_rejected)
+        self.outgoing_total = int(self.outgoing_total)
+        self.outgoing_advertised = int(self.outgoing_advertised)
```
why do you need to cast those into int if they are already ints?

adambaumeister replied:
When they get deserialized from XML, they are strings. Instead of handling it at deserialization, we do it here, this way it's impossible to create an instance of this dataclass where those values are incorrectly string ints ("1")

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    _title = "PAN-OS Objects"
+
+
+def dataclass_from_dict(device: Union[Panorama, Firewall], d: dict, class_type: Callable):
```
I suggest to give a more meaningful names here to ```d``` and to ```class_type```. 
any one word variable name isn't that good and doesn't tell us enough about that

adambaumeister replied:
Agreed, done

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    if device.serial:
+        d["hostid"] = device.serial
+
+    r = {}
```
same here, please give meaningful name

adambaumeister replied:
Done

adambaumeister replied:
I had another pass at this, I think I got it all now, there were only a couple more cases of this I could see. I'll make sure in the future to use better naming conventions (that's what I get for doing so much golang recently...)

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    return class_type(**r)
+
+
+def flatten_xml_to_dict(element: Element, d: dict, class_type: Callable):
```
I suggest to give a more meaningful names here to d and to class_type.
any one word variable name isn't that good and doesn't tell us enough about that

adambaumeister replied:
Done, though i think class_type is fine

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    Turns an XML `Element` Object into an instance of the provided dataclass. Dataclass paramaters must match
+    child XML tags exactly.
+    """
+    d = {}
```
I suggest to give a more meaningful names here to d and to class_type.
any one word variable name isn't that good and doesn't tell us enough about that

adambaumeister replied:
Done

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    child XML tags exactly.
+    """
+    d = {}
+    if element is None:
```
```suggestion
    if not element:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    if device.serial:
+        d["hostid"] = device.serial
+
+    child_element: Element
```
i might be missing something, but i don't see it being used anywhere

adambaumeister replied:
true - removed

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    """
+    Gets the name of a given PanDevice container or if it's not a container, returns shared.
+    """
+    if type(container) in [Panorama, Firewall]:
```
```suggestion
    if isinstance(container, (Panorama, Firewall)):
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+        device: Panorama
+        result = []
+        for device in topology.active_top_level_devices(device_filter_str):
+            if type(device) is Panorama:
```
```suggestion
            if isinstance(device, Panorama):
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+                for device_group_xml in response.findall("./result/devicegroups/entry"):
+                    dg_name = get_element_attribute(device_group_xml, "name")
+                    for device_xml in device_group_xml.findall("./devices/entry"):
+                        r: DeviceGroupInformation = dataclass_from_element(device, DeviceGroupInformation,
```
please give meaningful variable name

adambaumeister replied:
done

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    def get_device_groups(topology: Topology, device_filter_str: str = None) -> List[DeviceGroupInformation]:
+        """Get all the device groups from Panorama and their associated devices."""
+        device: Panorama
+        result = []
```
why not just call this variable device_groups?

adambaumeister replied:
Easier to name this result, as it's the result of the command. Makes it easier to copy and paste for new functions as well :)

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+        """Get all the template-stacks from Panorama and their associated devices."""
+
+        device: Panorama
+        result = []
```
why not just call this variable template_stacks?

adambaumeister replied:
Easier to name this result, as it's the result of the command. Makes it easier to copy and paste for new functions as well :)

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+                for template_stack_xml in response.findall("./result/template-stack/entry"):
+                    template_name = get_element_attribute(template_stack_xml, "name")
+                    for device_xml in template_stack_xml.findall("./devices/entry"):
+                        r: TemplateStackInformation = dataclass_from_element(device, TemplateStackInformation,
```
please give meaningful variable name

adambaumeister replied:
This, and other occurences of this, should be fixed now.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    def push_all(
+            topology: Topology,
+            device_filter_str: str = None,
+            device_group_filter: List[str] = None,
```
```suggestion
            device_group_filter: Optional[List[str]] = None,
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+            topology: Topology,
+            device_filter_str: str = None,
+            device_group_filter: List[str] = None,
+            template_stack_filter: List[str] = None
```
```suggestion
            template_stack_filter:  Optional[List[str]] = None
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    ) -> List[PushStatus]:
+        """Pushes the pending configuration from Panorama to the firewalls. This is an async function,
+        and will only push if there is config pending."""
+        result = []
```
why not call it push_statuses or something similar?

adambaumeister replied:
Easier to name this result, as it's the result of the command. Makes it easier to copy and paste for new functions as well :)

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+                    style="device group",
+                    name=dg_name
+                )
+                result_job_id = device.commit(cmd=device_group_commit)  # type: ignore
```
try to avoid ```# type: ignore``` as much as possible. if you can't avoid that lets look together

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    @staticmethod
+    def get_push_status(topology: Topology, match_job_ids: List[str] = None) -> List[PushStatus]:
+        """Retrieves the status of a Panorama Push, using the given job ids."""
+        result: List[PushStatus] = []
```
why not call it push_statuses or something similar?

adambaumeister replied:
See above comments about the use of result as a variable name

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+
+    @staticmethod
+    def get_available_software(topology: Topology,
+                               device_filter_str: str = None) -> SoftwareVersionCommandResult:
```
```suggestion
                               device_filter_str: Optional[str] = None) -> SoftwareVersionCommandResult:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+                    status = find_text_in_element(job, "./status")
+                    job_id = find_text_in_element(job, "./id")
+                    commit_type = find_text_in_element(job, "./type")
+                    if type(device) is Panorama:
```
```suggestion
                    if isinstance(device, Panorama):
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+
+    @staticmethod
+    def get_commit_job_status(topology: Topology, match_job_ids: List[str] = None) -> List[CommitStatus]:
+        result: List[CommitStatus] = []
```
could its name be set to commit_statuses?

adambaumeister replied:
Easier to name this result, as it's the result of the command. Makes it easier to copy and paste for new functions as well :)

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    REQUEST_STATE_PREFIX = "request high-availability state"
+
+    @staticmethod
+    def get_arp_table(topology: Topology, device_filter_str: str = None) -> ShowArpCommandResult:
```
```suggestion
    def get_arp_table(topology: Topology, device_filter_str: Optional[str] = None) -> ShowArpCommandResult:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+
+    @staticmethod
+    def get_counter_global(topology: Topology,
+                           device_filter_str: str = None) -> ShowCounterGlobalCommmandResult:
```
```suggestion
                           device_filter_str: Optional[str] = None) -> ShowCounterGlobalCommmandResult:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+
+    @staticmethod
+    def get_routing_summary(topology: Topology,
+                            device_filter_str: str = None) -> ShowRouteSummaryCommandResult:
```
```suggestion
                            device_filter_str: Optional[str] = None) -> ShowRouteSummaryCommandResult:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+
+    @staticmethod
+    def get_ha_status(topology: Topology,
+                      device_filter_str: str = None) -> List[ShowHAState]:
```
```suggestion
                      device_filter_str: Optional[str] = None) -> List[ShowHAState]:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+
+    @staticmethod
+    def get_routes(topology: Topology,
+                   device_filter_str: str = None) -> ShowRoutingRouteCommandResult:
```
```suggestion
                   device_filter_str: Optional[str] = None) -> ShowRoutingRouteCommandResult:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    :param topology: `Topology` instance !no-auto-argument
+    :param device_filter_string: String to filter to only show specific hostnames or serial numbers.
+    """
+    result: ShowSystemInfoCommandResult = UniversalCommand.get_system_info(topology,
+                                                                           device_filter_string)
+    return result
```
```suggestion
    return UniversalCommand.get_system_info(topology, device_filter_string)
```

GuyAfik replied:
You can do the same for all the functions below as well

adambaumeister replied:
Done

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    if not hasattr(result, "summary_data"):
+        # If this isn't a regular summary/result return, but instead, is just one object or a list of flat
+        # objects
+        if type(result) is list:
```
```suggestion
        if isinstance(result, list):
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7613,6 +9640,31 @@ def main():
             panorama_install_file_content_update_command(args)
 
+        elif demisto.command() == 'pan-os-platform-get-arp-tables':
+            topology = get_topology()
```
why do we need to get topology on each one of the commands? use it before running all the if statements

adambaumeister replied:
We don't need the setup costs of the topology object when a non-topology (legacy) command is being called. this saves a few API calls and makes the code more efficient.

GuyAfik replied:
then i suggest to end the if's of the panorama commands and then start a new series of if statement with pan-os commands while you call this line ```topology = get_topology()```

adambaumeister replied:
Can you think of a nice way to do this? If we  do it this way we will always create a topology object unless we add return to every other elif blocks. This is why I did it with the decorators, by the way, originally!

adambaumeister replied:
But this will cause get_topology() to be evaluated every time, even if the command above it already matched. For example if the command is pan-os-install-file-content-update, then this command will run then topology will be created and no further action will be taken

GuyAfik replied:
you are right, so just use it inside the if statement, i guess if would give us better grasp of what command is based on topology object.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7613,6 +9640,31 @@ def main():
+        elif demisto.command() == 'pan-os-platform-get-arp-tables':
+            topology = get_topology()
+            result = get_arp_tables(topology, **demisto.args())
+            dataclasses_to_command_results(result)
```
call dataclasses_to_command_results function inside the command function (in this case get_arp_tables)

adambaumeister replied:
This breaks the design of those functions. We want them to return the dataclass instances, not be responsible for the interface to XSOAR. This should make this code much more maintainable in the future.

GuyAfik commented:
Hi Adam, incredible work has been done! a few points:

1) To some of the comments of the previous review, I don't see you answered or changed, if there is any problem or we need to discuss further, feel free to reach me
2) Please add unit-tests for the entire functionality that you have added if that's a possibility.
3) We will need to add the new commands being add in this PR to a test-playbook, lets discuss further.
4) Please try to avoid referring to the type of the variable when creating it.

For any questions/problem/assistance feel free to nudge me at any time.

GuyAfik commented:
> I updated to fix all these other issues, and I resolved the conflicts via a merge. Please double check the merge worked correctly.
> 
> I can't get the build to succeed though - there are several MyPy errors I'm not sure how to resolve, if you can help?
> 
> I also added new release notes, 1.9.0, as that's the current version? update-release-notes did not work for me here, I had to add them manually for some reason.

You have resolved the merge conflicts good, regarding the build, once your PR is ready ill take a look at the MyPy errors, for now just ignore it. Release note is valid as well. Great work!

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -4564,58 +4580,322 @@ script:
+    - contextPath: PANOS.ShowArp.Summary.max
+      description: Maximum supported ARP Entries
+      type: Unknown
+    - contextPath: PANOS.ShowArp.Summary.total
```
Done

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -4564,58 +4580,322 @@ script:
+      type: Unknown
+  - arguments:
+    - default: false
+      defaultValue: null
```
done

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+
+    result = func(topology, **demisto_args)
+    if not result:
+        command_result = CommandResults(
```
Good idea. I updated the dataclass converter metthod to allow passing of a "no results" string. This keeps the rest of the code simple and makes it clear.

GuyAfik replied:
great work!

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+    api_key: str = ""
+
+
+def convert_params_to_object(params: dict) -> DemistoParameters:
```
done

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7046,8 +7062,10 @@ def prettify_configured_user_id_agents(user_id_agents: Union[List, Dict]) -> Uni
             'Name': user_id_agents['@name'],
             'Host': dict_safe_get(user_id_agents, keys=['host-port', 'host']),
             'Port': dict_safe_get(user_id_agents, keys=['host-port', 'port']),
-            'NtlmAuth': dict_safe_get(user_id_agents, keys=['host-port', 'ntlm-auth'], default_return_value='no'),
-            'LdapProxy': dict_safe_get(user_id_agents, keys=['host-port', 'ldap-proxy'], default_return_value='no'),
+            'NtlmAuth': dict_safe_get(user_id_agents, keys=['host-port', 'ntlm-auth'],
+                                      default_return_value='no'),
+            'LdapProxy': dict_safe_get(user_id_agents, keys=['host-port', 'ldap-proxy'],
+                                       default_return_value='no'),
```
Done

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+
+
+# pan-os-python new classes
+class CustomVersionedPanObject(VersionedPanObject):
```
added

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+        if xml is None:
+            xml = self._refresh_xml(running_config, True)
+
+        if xml is None:
```
Fixed

adambaumeister replied:
Not fixed, actually;

So - these have to be like this due to the way xml.etree.Element objects are nullable.

It seems like if an XML element has no text (???) but has child elements, it will return false  when used in a conditional. So, we need to check it's literally a NoneType object.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+        child = self.CHILDTYPES[0]()
+        child.parent = self
+        childroot = xml.find(child.XPATH[1:])
+        if childroot is not None:
```
Fixed

GuyAfik replied:
actually that is not is not correct, do you need here all the childroot which have a some kind of value that isn't None? if yes just use ```if childroot:```

adambaumeister replied:
maybe lets discuss this one on a call, it's strange the way elementree handles it. See this stackoverflow answer that implies the way I've got it is the objectively correct way (that is, check if the element is literally type `None`)

https://stackoverflow.com/questions/20129996/why-does-boolxml-etree-elementtree-element-evaluate-to-false

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+
+
+def run_op_command(device: Union[Panorama, Firewall], cmd: str, **kwargs) -> Element:
+    result: Element = device.op(cmd, **kwargs)
```
changed

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+
+def find_text_in_element(element: Element, tag: str) -> str:
+    result = element.find(tag)
+    if result is None:
```
fixed

adambaumeister replied:
Just to confirm - see the above comment, this is the correct way to handle the case of an empty reuslt Element

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    if not hasattr(result, "text"):
+        raise LookupError(f"Tag {tag} has no text.")
+
+    if result.text:
+        return result.text
+    else:
+        return ""
```
yes, if the xml element only has child elements I think it will be none.

adambaumeister replied:
fixed

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7265,2013 @@ def panorama_install_file_content_update_command(args: dict):
+    def _refresh_children(self, running_config=False, xml=None):
+        """Override normal refresh method"""
+        # Retrieve the xml if we weren't given it
+        if xml is None:
```
So - these have to be like this due to the way xml.etree.Element objects are nullable.

It seems like if an XML element has no text (???) but has child elements, it will return false  when used in a conditional. So, we need to check it's literally a NoneType object.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+        )
+
+    @staticmethod
+    def get_counter_global(topology: Topology, device_filter_str: str = None) -> ShowCounterGlobalCommmandResult:
```
```suggestion
    def get_counter_global(topology: Topology, device_filter_str: Optional[str] = None) -> ShowCounterGlobalCommmandResult:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+        )
+
+    @staticmethod
+    def get_routing_summary(topology: Topology, device_filter_str: str = None) -> ShowRouteSummaryCommandResult:
```
```suggestion
    def get_routing_summary(topology: Topology, device_filter_str: Optional[str] = None) -> ShowRouteSummaryCommandResult:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+        )
+
+    @staticmethod
+    def get_bgp_peers(topology: Topology, device_filter_str: str = None) -> ShowRoutingProtocolBGPCommandResult:
```
```suggestion
    def get_bgp_peers(topology: Topology, device_filter_str: Optional[str] = None) -> ShowRoutingProtocolBGPCommandResult:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+            return direct_firewall_connection.xapi.export_result.get("content")
+
+    @staticmethod
+    def get_ha_status(topology: Topology, device_filter_str: str = None) -> List[ShowHAState]:
```
```suggestion
    def get_ha_status(topology: Topology, device_filter_str: Optional[str] = None) -> List[ShowHAState]:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+        )
+
+    @staticmethod
+    def get_routes(topology: Topology, device_filter_str: str = None) -> ShowRoutingRouteCommandResult:
```
```suggestion
    def get_routes(topology: Topology, device_filter_str: Optional[str] = None) -> ShowRoutingRouteCommandResult:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7613,6 +9603,30 @@ def main():
+            dataclasses_to_command_results(result)
+        elif demisto.command() == 'pan-os-platform-get-device-groups':
+            topology = get_topology()
+            result = get_device_groups(topology, **demisto.args())
+            dataclasses_to_command_results(result, empty_result_message="No device groups found.")
```
```suggestion
            dataclasses_to_command_results(get_device_groups(topology, **demisto.args()), empty_result_message="No device groups found.")
```

GuyAfik replied:
please change that to all the commands

adambaumeister replied:
I changed this to wrap dataclasses_to_command_results with return_results instead, I think this is the most readable. let me know.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+    return "ok"
+
+
+def get_arp_tables(topology: Topology, device_filter_string: str = None) -> ShowArpCommandResult:
```
```suggestion
def get_arp_tables(topology: Topology, device_filter_string: Optional[str] = None) -> ShowArpCommandResult:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+
+
+def get_route_summaries(topology: Topology,
+                        device_filter_string: str = None) -> ShowRouteSummaryCommandResult:
```
```suggestion
                        device_filter_string: Optional[str] = None) -> ShowRouteSummaryCommandResult:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+
+
+def get_routes(topology: Topology,
+               device_filter_string: str = None) -> ShowRoutingRouteCommandResult:
```
```suggestion
               device_filter_string: Optional[str] = None) -> ShowRoutingRouteCommandResult:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+
+
+def get_system_info(topology: Topology,
+                    device_filter_string: str = None) -> ShowSystemInfoCommandResult:
```
```suggestion
                    device_filter_string: Optional[str] = None) -> ShowSystemInfoCommandResult:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+    return result
+
+
+def get_device_groups(topology: Topology, device_filter_string: str = None) -> List[DeviceGroupInformation]:
```
```suggestion
def get_device_groups(topology: Topology, device_filter_string: Optional[str] = None) -> List[DeviceGroupInformation]:
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+    return result
+
+
+def get_template_stacks(topology: Topology, device_filter_string: str = None) -> List[
```
```suggestion
def get_template_stacks(topology: Topology, device_filter_string: Optional[str] = None) -> List[
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+    server_url = demisto.params().get('server')
+    parsed_url = urlparse(server_url)
+    hostname = parsed_url.hostname
+    topology = Topology.build_from_string(
+        hostname,
+        username="",
+        password="",
+        api_key=demisto.params().get('key')
+    )
+    return topology
```
```suggestion
    return Topology.build_from_string(
        hostname,
        username="",
        password="",
        api_key=demisto.params().get('key')
    )
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+        command_result = CommandResults(
+            readable_output=empty_result_message,
+        )
+        return_results(command_result)
+        return command_result
```
why do you use return_results and then return the command_result from the function? This is redundant

GuyAfik replied:
I suggest to return the command_result from the function and then just call return_results in the main

adambaumeister replied:
I'm not sure why I did it this way originally - I've removed that.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+        **extra_args
+    )
+    return_results(command_result)
+    return command_result
```
why do you use return_results and then return the command_result from the function? This is redundant

GuyAfik replied:
I suggest to return the command_result from the function and then just call return_results in the main

adambaumeister replied:
Agreed - fixed

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7261,7 +9249,9 @@ def main():
 
         if demisto.command() == 'test-module':
             panorama_test()
-
+            # Test the topology is working as well
+            topology = get_topology()
+            test_topology_connectivity(topology)
```
This still needs to be fixed. I still think we should have a 'test_module' function where both panorama_test and test_topology_connectivity are called. could there be a customer which uses panorama commands but won't use the topology? if yes lets discuss

adambaumeister replied:
Yes, there would definitely be customers that don't need the topology commands (and vice versa).

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+
+
+class AntiSpywareProfileBotnetDomainList(CustomVersionedPanObject):
+    ROOT = Root.VSYS  # pylint: disable=E1101
```
Cant' remember the cause of this, it has to do with instance member not having an attirbute because it's set dynamically. I have removed it.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+    @staticmethod
+    def get_device_groups(topology: Topology, device_filter_str: str = None) -> List[DeviceGroupInformation]:
+        """Get all the device groups from Panorama and their associated devices."""
+        device: Panorama
```
Fixed

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+        TemplateStackInformation]:
+        """Get all the template-stacks from Panorama and their associated devices."""
+
+        device: Panorama
```
Fixed

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+    """
+    host_id: str = ""
+    if device.hostname:
+        host_id = device.hostname
```
It could be both but the logic is to use the serial instead of the hostname if it exists.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+                    hostid=resolve_host_id(device),
+                    commit_type="devicegroup",
+                    name=dg_name,
+                    job_id=result_job_id,  # type: ignore
```
Done

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+                    style="template stack",
+                    name=template_name
+                )
+                result_job_id = device.commit(cmd=template_stack_commit)  # type: ignore
```
Done

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+                    hostid=resolve_host_id(device),
+                    commit_type="template-stack",
+                    name=template_name,
+                    job_id=result_job_id,  # type: ignore
```
Done

adambaumeister commented:
Heya Guy

As per our last discussion, I merged the testing functions and wrote unit tests for all the commands as well as the Topology class. Please let me know if this unit test structure is OK.

I had a look at the type: ignore lines, I can't see anything else we can do with them, but if you want to remove them (and ignore the mypy errors) this is ok by me as well.

Cheers
Adam

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7271,1976 @@ def panorama_install_file_content_update_command(args: dict):
+        object_dict["hostid"] = device.serial
+
+    r = {}
+    for k, v in object_dict.items():
```
what is k and what is v here? please give meaningful names

GuyAfik commented:
> Heya Guy
> 
> As per our last discussion, I merged the testing functions and wrote unit tests for all the commands as well as the Topology class. Please let me know if this unit test structure is OK.
> 
> I had a look at the type: ignore lines, I can't see anything else we can do with them, but if you want to remove them (and ignore the mypy errors) this is ok by me as well.
> 
> Cheers Adam

Hi Adam,
Nice work with the unit tests. The structure is great. see my minor comments in the unit-tests file.
regarding the ignore types, leave it as it is, I will deal with that later on if required.

In addition please make sure you have done the following:
1) add doc-string to every function that you created/added following this pattern if possible - https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html. I see to some functions you did add a doc-string and to some you didn't. try to be consistent.
2) Some of the issues where you wrote ```Done``` i don't see a change has been actually done, please take a look.
3) add docstrings to unit-test to get a better understanding of what they test.
4) go through the minor comments i gave you, some of them are related to what I just wrote here.

Overall great work, not much left to fix.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama_test.py:
```diff
@@ -1022,3 +1026,308 @@ def test_without_specify(self, requests_mock):
+
+    @patch("Panorama.Topology.get_all_child_firewalls")
+    @patch("Panorama.run_op_command")
+    def test_add_firewall_device_object(self, patched_run_op_command, _, mock_firewall):
```
@patch decorator passes the patched object as a function argument - if there's a way to prevent this behavior we can remove these _ arguments  in the test functions that use them.

GuyAfik replied:
not sure how, need to check

GuyAfik replied:
not sure how, need to check

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama_test.py:
```diff
@@ -1022,3 +1026,308 @@ def test_without_specify(self, requests_mock):
+
+    @patch("Panorama.Topology.get_all_child_firewalls")
+    @patch("Panorama.run_op_command")
+    def test_add_panorama_device_object(self, patched_run_op_command, _, mock_panorama):
```
I've added it but tried to keep it short and sweet, let me know if you want me to expand on it further as you've described here

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7271,1976 @@ def panorama_install_file_content_update_command(args: dict):
+    :param topology: `Topology` instance !no-auto-argument
+    :param device_filter_string: String to filter to only show specific hostnames or serial numbers.
+    """
+    result: ShowRouteSummaryCommandResult = FirewallCommand.get_routing_summary(topology, device_filter_string)
```
Done

adambaumeister commented:
Hey Guy,

I hope I found all the stuff that was still outstanding. It's getting super hard to parse all the comments in this PR, I think I went through everything that wasn't outdated or resolved. I also updated the docstrings as discussed.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7274,1965 @@ def panorama_install_file_content_update_command(args: dict):
+
+    def firewall_devices(self) -> ValuesView[Firewall]:
+        """
+        Retunrs the firewall devices in the topology
```
```suggestion
        Returns the firewall devices in the topology
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7274,1965 @@ def panorama_install_file_content_update_command(args: dict):
+            api_password=password,
+        )
+        # Set the timeout
+        device.timeout = 120
```
```suggestion
        device.timeout = DEVICE_TIMEOUT
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7274,1965 @@ def panorama_install_file_content_update_command(args: dict):
+
+        result = []
+        for device in topology.active_devices(device_filter_string):
+            job_type = "Commit"
```
what is it necessary to define that variable inside the for-loop? can't it be outside the for-loop? if its just used once in the function, then there isn't a reason to keep that variable, you can just remove it and do ```commit_type='Commit'```

adambaumeister replied:
Agreed, fixed

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7274,1965 @@ def panorama_install_file_content_update_command(args: dict):
+    @staticmethod
+    def change_status(topology: Topology, hostid: str, state: str) -> HighAvailabilityStateStatus:
+        firewall = list(topology.firewalls(filter_string=hostid))[0]
+        run_op_command(firewall, FirewallCommand.REQUEST_STATE_PREFIX + f" {state}")
```
```suggestion
        run_op_command(firewall, f'{FirewallCommand.REQUEST_STATE_PREFIX} {state}')
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7274,1965 @@ def panorama_install_file_content_update_command(args: dict):
+    :param empty_result_message: If the result data is non
+    """
+    if not result:
+        command_result = CommandResults(
+            readable_output=empty_result_message,
+        )
+        return_results(command_result)
+        return command_result
```
```suggestion
        return CommandResults(
            readable_output=empty_result_message,
        )
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7271,1976 @@ def panorama_install_file_content_update_command(args: dict):
+            device_retrieval_func = self.active_top_level_devices
+
+        for device in device_retrieval_func(device_filter_string):
+            print(device)
```
can you remove that please? maybe you meant to log here something?

adambaumeister replied:
yep, removed this

GuyAfik commented:
Hi @adambaumeister, nice work! I think this week we can do a short demo as we are almost done.

Can you add doc-string to every function that you created/added following this pattern if possible - https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html. I see to some functions you did add a doc-string and to some you didn't. 

In addition refer to my new very minor review above and to all unresolved threads if that's possible.

ShirleyDenkberg commented:
@adambaumeister On lines Lines 6930 - 6935 - Should these types be a string or a number?
I added the new commands to the list of commands in the beginning of the README.  I also fixed all the links so that they now work.
@GuyAfik @bakatzir Doc review completed.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7263,1978 @@ def panorama_install_file_content_update_command(args: dict):
+                    if isinstance(container[1], Panorama):
+                        return_containers.append(container)
+                if not isinstance(container[1], (Panorama, Firewall)):
+                    if container[1].name == container_name:  # type: ignore
```
Removed

GuyAfik commented on Packs/PAN-OS/ReleaseNotes/1_10_0.md:
```diff
@@ -0,0 +1,11 @@
+  - ***pan-os-platform-get-system-info***
+  - ***pan-os-platform-get-device-groups***
+  - ***pan-os-platform-get-template-stacks***
+- Updated the Docker image to: *demisto/pan-os-python*.
```
```suggestion
- Updated the Docker image to: *demisto/pan-os-python:1.0.0.27504*.
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7247,6 +7274,2082 @@ def panorama_install_file_content_update_command(args: dict):
+    if hasattr(result, "_outputs_key_field"):
+        extra_args["outputs_key_field"] = getattr(result, "_outputs_key_field")
+
+    readable_output = tableToMarkdown(title, summary_list)
```
```suggestion
    readable_output = tableToMarkdown(title, summary_list, removeNull=True)
```

That will remove automatically empty columns if exist

GuyAfik commented:
Very nice work @adambaumeister, Thank you for the massive efforts that you made. 
The PR looks great.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=17689) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=17689) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: adambaumeister<br/>:x: erezh31<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=17689) it.</sub>
